File: frontend/src/types/recruitment.ts
/**
 * Stage 4: Recruitment & Interview Intelligence TypeScript Definitions
 */

export interface JobRequirementSkillItem {
  skill_id: string;
  skill_name: string;
  min_proficiency: number;
  weight: number;
  importance: "required" | "preferred";
}

export interface JobRequirementWeights {
  required_skills: number;
  preferred_skills: number;
  evidence_strength: number;
  experience_alignment: number;
}

export interface JobRequirementVersion {
  id: string;
  job_opening_id: string;
  version_number: number;
  responsibilities: string;
  required_skills: JobRequirementSkillItem[];
  preferred_skills: JobRequirementSkillItem[];
  min_years_experience: number;
  weights: JobRequirementWeights;
  quality_audit_flags: string[];
  is_active: boolean;
  created_by: string;
  created_at: string;
}

export interface JobOpening {
  id: string;
  organization_id: string;
  department_id: string;
  department_name?: string | null;
  job_role_id: string;
  job_role_title?: string | null;
  requisition_code: string;
  title: string;
  location: string;
  employment_type: string;
  target_headcount: number;
  hiring_manager_profile_id?: string | null;
  recruiter_profile_id?: string | null;
  status: "active" | "draft" | "paused" | "closed" | "fulfilled";
  current_requirement_version: number;
  active_requirement?: JobRequirementVersion | null;
  candidate_count: number;
  created_at: string;
  updated_at: string;
}

export interface JobOpeningCreatePayload {
  department_id: string;
  job_role_id: string;
  requisition_code: string;
  title: string;
  location?: string;
  employment_type?: string;
  target_headcount?: number;
  responsibilities: string;
  required_skills: JobRequirementSkillItem[];
  preferred_skills?: JobRequirementSkillItem[];
  min_years_experience?: number;
  weights?: Partial<JobRequirementWeights>;
}

export interface ResumeDocument {
  id: string;
  organization_id: string;
  candidate_id: string;
  candidate_name?: string | null;
  job_opening_id?: string | null;
  file_name: string;
  file_size_bytes: number;
  mime_type: string;
  sha256_hash: string;
  page_count: number;
  parser_name: string;
  is_ocr_required: boolean;
  parsing_status: "uploaded" | "parsing" | "extracted" | "ocr_required" | "failed";
  error_message?: string | null;
  uploaded_at: string;
}

export interface ExtractedSkillItem {
  name: string;
  category?: string | null;
  source_context: string;
  page_number?: number | null;
}

export interface WorkExperienceEntry {
  company: string;
  title: string;
  start_date?: string | null;
  end_date?: string | null;
  is_current: boolean;
  responsibilities: string[];
  demonstrated_skills: string[];
}

export interface ProjectEntry {
  name: string;
  description: string;
  technologies: string[];
  role?: string | null;
  link?: string | null;
}

export interface ResumeExtraction {
  id: string;
  resume_id: string;
  candidate_id: string;
  model_version: string;
  prompt_version: string;
  summary: string;
  work_experiences: WorkExperienceEntry[];
  projects: ProjectEntry[];
  education: Array<Record<string, unknown>>;
  certifications: string[];
  extracted_skills: ExtractedSkillItem[];
  total_years_experience: number;
  extraction_status: "raw_extracted" | "verified" | "rejected";
  reviewed_by?: string | null;
  created_at: string;
}

export interface SkillNormalizationItem {
  id: string;
  organization_id: string;
  candidate_id: string;
  raw_phrase: string;
  canonical_skill_id?: string | null;
  canonical_skill_name?: string | null;
  canonical_skill_category?: string | null;
  match_method: "exact_canonical" | "exact_alias" | "graph_adjacent" | "controlled_fuzzy" | "ambiguous" | "unresolved";
  confidence: number;
  review_status: "pending" | "confirmed" | "overridden" | "rejected";
  reviewed_by?: string | null;
  created_at: string;
}

export interface CriterionScoreDetail {
  criterion_name: string;
  weight: number;
  raw_score: number;
  weighted_contribution: number;
  evidence_citations: string[];
  notes?: string | null;
}

export interface CandidateMatchEvaluation {
  id: string;
  job_opening_id: string;
  requirement_version_id: string;
  candidate_id: string;
  candidate_name: string;
  candidate_email: string;
  rank_position: number;
  overall_match_score: number;
  required_skill_coverage: number;
  preferred_skill_coverage: number;
  evidence_strength_score: number;
  experience_alignment_score: number;
  criterion_breakdown: Record<string, CriterionScoreDetail>;
  qwen_explanation?: string | null;
  explanation_grounding_status: "grounded" | "unverified" | "failed_grounding" | "degraded_mode";
  is_stale: boolean;
  calculated_at: string;
}

export interface CandidateRankingListResponse {
  job_opening_id: string;
  job_title: string;
  requirement_version: number;
  candidates: CandidateMatchEvaluation[];
  total_evaluated: number;
}

export interface RubricLevel {
  level: number;
  title: string;
  description: string;
}

export interface InterviewQuestionItem {
  question_index: number;
  competency: string;
  question_type: "opening" | "competency" | "situational" | "evidence_validation" | "follow_up";
  question_text: string;
  why_asking: string;
  expected_evidence: string;
  rubric: RubricLevel[];
  follow_up_prompts: string[];
  interviewer_notes?: string | null;
}

export interface InterviewKit {
  id: string;
  organization_id: string;
  job_opening_id: string;
  candidate_id?: string | null;
  candidate_name?: string | null;
  title: string;
  stage: string;
  questions: InterviewQuestionItem[];
  status: "draft" | "approved" | "archived";
  approved_by?: string | null;
  approved_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface InterviewResponseRecord {
  id: string;
  question_index: number;
  question_text: string;
  competency: string;
  candidate_response_text: string;
  interviewer_notes?: string | null;
  recorded_at: string;
}

export interface InterviewSession {
  id: string;
  organization_id: string;
  interview_kit_id: string;
  candidate_id: string;
  candidate_name: string;
  interviewer_profile_id: string;
  interviewer_name?: string | null;
  scheduled_at: string;
  session_status: "scheduled" | "in_progress" | "completed" | "cancelled";
  completed_at?: string | null;
  notes?: string | null;
  responses: InterviewResponseRecord[];
  created_at: string;
}

export interface RubricAssessmentItem {
  competency: string;
  assessed_level: number;
  evidence_demonstrated: string;
  evidence_missing?: string | null;
}

export interface InterviewInsight {
  id: string;
  interview_session_id: string;
  candidate_id: string;
  summary: string;
  rubric_analysis: RubricAssessmentItem[];
  demonstrated_strengths: string[];
  evidence_gaps: string[];
  confidence_band: "high" | "moderate" | "limited" | "insufficient";
  follow_up_recommendations: string[];
  review_status: "ai_analyzed" | "human_approved" | "human_overridden";
  reviewed_by?: string | null;
  created_at: string;
}

export interface RecruitmentDecision {
  id: string;
  organization_id: string;
  job_opening_id: string;
  candidate_id: string;
  decided_by: string;
  decided_by_name?: string | null;
  decision: "advance" | "shortlist" | "hold" | "request_info" | "schedule_interview" | "offer" | "reject" | "withdrawn";
  ai_recommendation?: string | null;
  is_override: boolean;
  override_reason?: string | null;
  rationale: string;
  candidate_facing_status: string;
  created_at: string;
}

export interface MultiBrainRun {
  id: string;
  organization_id: string;
  role_name: string;
  model_name: string;
  prompt_version: string;
  input_ref?: string | null;
  output_summary?: string | null;
  duration_ms: number;
  status: "completed" | "failed" | "degraded_fallback";
  created_at: string;
}

export interface CandidateApplication {
  application_id: string;
  job_id?: string | null;
  job_title: string;
  location: string;
  status: string;
  applied_at: string;
}

File: backend/app/schemas/recruitment.py
"""Pydantic schemas for Stage 4: Recruitment and Interview Intelligence."""

from typing import Any, Dict, List, Literal, Optional, Union
from pydantic import BaseModel, Field


# ====================================================================
# 1. Job Opening & Requirements Schemas
# ====================================================================

class JobRequirementSkillItem(BaseModel):
    skill_id: str
    skill_name: str
    min_proficiency: int = Field(default=1, ge=1, le=5)
    weight: float = Field(default=1.0, ge=0.0, le=1.0)
    importance: Literal["required", "preferred"] = "required"


class JobRequirementWeights(BaseModel):
    required_skills: float = Field(default=0.45, ge=0.0, le=1.0)
    preferred_skills: float = Field(default=0.20, ge=0.0, le=1.0)
    evidence_strength: float = Field(default=0.20, ge=0.0, le=1.0)
    experience_alignment: float = Field(default=0.15, ge=0.0, le=1.0)


class JobOpeningCreate(BaseModel):
    department_id: str
    job_role_id: str
    requisition_code: str = Field(..., min_length=3, max_length=50)
    title: str = Field(..., min_length=3, max_length=150)
    location: str = "Remote"
    employment_type: Literal["full_time", "part_time", "contract"] = "full_time"
    target_headcount: int = Field(default=1, ge=1)
    responsibilities: str = Field(..., min_length=10)
    required_skills: List[JobRequirementSkillItem] = Field(default_factory=list)
    preferred_skills: List[JobRequirementSkillItem] = Field(default_factory=list)
    min_years_experience: float = Field(default=0.0, ge=0.0)
    weights: JobRequirementWeights = Field(default_factory=JobRequirementWeights)


class JobRequirementVersionResponse(BaseModel):
    id: str
    job_opening_id: str
    version_number: int
    responsibilities: str
    required_skills: List[JobRequirementSkillItem]
    preferred_skills: List[JobRequirementSkillItem]
    min_years_experience: float
    weights: Dict[str, float]
    quality_audit_flags: List[str]
    is_active: bool
    created_at: str


class JobOpeningResponse(BaseModel):
    id: str
    organization_id: str
    department_id: str
    department_name: Optional[str] = None
    job_role_id: str
    job_role_title: Optional[str] = None
    requisition_code: str
    title: str
    location: str
    employment_type: str
    target_headcount: int
    hiring_manager_profile_id: Optional[str] = None
    recruiter_profile_id: Optional[str] = None
    status: Literal["draft", "active", "paused", "closed"]
    current_requirement_version: int
    active_requirement: Optional[JobRequirementVersionResponse] = None
    candidate_count: int = 0
    created_at: str
    updated_at: str


# ====================================================================
# 2. Resume Ingestion & Parsing Schemas
# ====================================================================

class ResumeDocumentResponse(BaseModel):
    id: str
    organization_id: str
    candidate_id: str
    candidate_name: Optional[str] = None
    job_opening_id: Optional[str] = None
    file_name: str
    file_size_bytes: int
    mime_type: str
    sha256_hash: str
    page_count: int
    parser_name: str
    is_ocr_required: bool
    parsing_status: Literal["uploaded", "parsing", "extracted", "ocr_required", "failed"]
    error_message: Optional[str] = None
    uploaded_at: str


class WorkExperienceEntry(BaseModel):
    company: str
    title: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    is_current: bool = False
    responsibilities: List[str] = Field(default_factory=list)
    demonstrated_skills: List[str] = Field(default_factory=list)


class ProjectEntry(BaseModel):
    name: str
    description: str
    technologies: List[str] = Field(default_factory=list)
    role: Optional[str] = None
    link: Optional[str] = None


class ExtractedSkillItem(BaseModel):
    name: str
    category: Optional[str] = None
    source_context: str
    page_number: Optional[int] = 1


class ResumeExtractionDTO(BaseModel):
    """Schema expected from Qwen for structured resume evidence extraction."""
    summary: str
    work_experiences: List[WorkExperienceEntry] = Field(default_factory=list)
    projects: List[ProjectEntry] = Field(default_factory=list)
    education: List[Dict[str, Any]] = Field(default_factory=list)
    certifications: List[str] = Field(default_factory=list)
    extracted_skills: List[ExtractedSkillItem] = Field(default_factory=list)
    total_years_experience: float = 0.0


class ResumeExtractionResponse(BaseModel):
    id: str
    resume_id: str
    candidate_id: str
    model_version: str
    prompt_version: str
    summary: Optional[str] = None
    work_experiences: List[WorkExperienceEntry]
    projects: List[ProjectEntry]
    education: List[Dict[str, Any]]
    certifications: List[str]
    extracted_skills: List[ExtractedSkillItem]
    review_status: Literal["pending_review", "accepted", "corrected", "rejected"]
    reviewed_by: Optional[str] = None
    created_at: str


# ====================================================================
# 3. Skill Normalization Schemas
# ====================================================================

class SkillNormalizationItem(BaseModel):
    id: str
    raw_phrase: str
    canonical_skill_id: Optional[str] = None
    canonical_skill_name: Optional[str] = None
    canonical_skill_code: Optional[str] = None
    match_method: Literal["exact_canonical", "exact_alias", "graph_adjacent", "controlled_fuzzy", "ambiguous", "unresolved"]
    confidence: float
    review_status: Literal["pending", "confirmed", "overridden", "rejected"]
    created_at: str


class SkillNormalizationUpdateRequest(BaseModel):
    canonical_skill_id: str
    review_status: Literal["confirmed", "overridden", "rejected"] = "confirmed"


# ====================================================================
# 4. Transparent Match & Ranking Schemas
# ====================================================================

class CriterionScoreDetail(BaseModel):
    criterion_name: str
    weight: float
    raw_score: float  # 0.0 to 100.0
    weighted_contribution: float  # raw_score * weight
    evidence_citations: List[str] = Field(default_factory=list)
    notes: Optional[str] = None


class CandidateMatchEvaluationResponse(BaseModel):
    id: str
    job_opening_id: str
    requirement_version_id: str
    candidate_id: str
    candidate_name: str
    candidate_email: str
    rank_position: int
    overall_match_score: float  # 0 to 100
    required_skill_coverage: float  # 0 to 100
    preferred_skill_coverage: float  # 0 to 100
    evidence_strength_score: float  # 0 to 100
    experience_alignment_score: float  # 0 to 100
    criterion_breakdown: Dict[str, CriterionScoreDetail]
    qwen_explanation: Optional[str] = None
    explanation_grounding_status: Literal["grounded", "unverified", "failed_grounding", "degraded_mode"]
    is_stale: bool
    calculated_at: str


class CandidateRankingListResponse(BaseModel):
    job_opening_id: str
    job_title: str
    requirement_version: int
    candidates: List[CandidateMatchEvaluationResponse]
    total_evaluated: int


class CandidateExplanationDTO(BaseModel):
    """Schema expected from Qwen for ranking rationale synthesis."""
    comparative_rationale: str
    exact_skills_validated: List[str]
    adjacent_skills_credited: List[Dict[str, Any]] = Field(default_factory=list)
    missing_critical_skills: List[str] = Field(default_factory=list)
    evidence_citations: List[Union[str, Dict[str, Any]]] = Field(default_factory=list)
    confidence_band: Literal["high", "moderate", "limited", "insufficient"] = "moderate"
    needs_human_review: bool = True


# ====================================================================
# 5. Interview Kits & Rubrics Schemas
# ====================================================================

class RubricLevel(BaseModel):
    level: int  # 1 to 5
    title: str
    description: str


class InterviewQuestionItem(BaseModel):
    question_index: int
    competency: str
    question_type: Literal["opening", "competency", "situational", "evidence_validation", "follow_up"]
    question_text: str
    why_asking: str
    expected_evidence: str
    rubric: List[RubricLevel]
    follow_up_prompts: List[str] = Field(default_factory=list)
    interviewer_notes: Optional[str] = None


class InterviewKitCreate(BaseModel):
    job_opening_id: str
    candidate_id: Optional[str] = None
    title: str
    stage: str = "technical_round_1"
    focus_competencies: Optional[List[str]] = None


class InterviewKitUpdate(BaseModel):
    title: Optional[str] = None
    questions: List[InterviewQuestionItem]


class InterviewKitResponse(BaseModel):
    id: str
    organization_id: str
    job_opening_id: str
    candidate_id: Optional[str] = None
    candidate_name: Optional[str] = None
    title: str
    stage: str
    questions: List[InterviewQuestionItem]
    status: Literal["draft", "approved", "archived"]
    approved_by: Optional[str] = None
    approved_at: Optional[str] = None
    created_at: str
    updated_at: str


class InterviewKitDTO(BaseModel):
    """Schema expected from Qwen Interview Architect."""
    title: str
    stage: str
    questions: List[InterviewQuestionItem]


# ====================================================================
# 6. Interview Session & Response Schemas
# ====================================================================

class InterviewSessionCreate(BaseModel):
    interview_kit_id: str
    candidate_id: str
    scheduled_at: str
    notes: Optional[str] = None


class InterviewResponseRecordRequest(BaseModel):
    question_index: int
    question_text: str
    competency: str
    candidate_response_text: str
    interviewer_notes: Optional[str] = None


class InterviewResponseRecord(BaseModel):
    id: str
    question_index: int
    question_text: str
    competency: str
    candidate_response_text: str
    interviewer_notes: Optional[str] = None
    recorded_at: str


class InterviewSessionResponse(BaseModel):
    id: str
    organization_id: str
    interview_kit_id: str
    candidate_id: str
    candidate_name: str
    interviewer_profile_id: str
    interviewer_name: Optional[str] = None
    scheduled_at: str
    session_status: Literal["scheduled", "in_progress", "completed", "cancelled"]
    completed_at: Optional[str] = None
    notes: Optional[str] = None
    responses: List[InterviewResponseRecord] = Field(default_factory=list)
    created_at: str


# ====================================================================
# 7. Structured Interview Insights Schemas
# ====================================================================

class RubricAssessmentItem(BaseModel):
    competency: str
    assessed_level: int = Field(ge=1, le=5)
    evidence_demonstrated: str
    evidence_missing: Optional[str] = None


class InterviewInsightDTO(BaseModel):
    """Schema expected from Qwen Interview Evidence Analyst."""
    summary: str
    rubric_analysis: List[RubricAssessmentItem]
    demonstrated_strengths: List[str]
    evidence_gaps: List[str]
    confidence_band: Literal["high", "moderate", "limited", "insufficient"]
    follow_up_recommendations: List[str]


class InterviewInsightResponse(BaseModel):
    id: str
    interview_session_id: str
    candidate_id: str
    summary: str
    rubric_analysis: List[RubricAssessmentItem]
    demonstrated_strengths: List[str]
    evidence_gaps: List[str]
    confidence_band: Literal["high", "moderate", "limited", "insufficient"]
    follow_up_recommendations: List[str]
    review_status: Literal["ai_analyzed", "human_approved", "human_overridden"]
    reviewed_by: Optional[str] = None
    created_at: str


# ====================================================================
# 8. Human Decision Gate Schemas
# ====================================================================

class RecruitmentDecisionCreate(BaseModel):
    decision: Literal["advance", "shortlist", "hold", "request_info", "schedule_interview", "offer", "reject", "withdrawn"]
    rationale: str = Field(..., min_length=5)
    is_override: bool = False
    override_reason: Optional[str] = None
    candidate_facing_status: str = "Under Review"


class RecruitmentDecisionResponse(BaseModel):
    id: str
    organization_id: str
    job_opening_id: str
    candidate_id: str
    decided_by: str
    decided_by_name: Optional[str] = None
    decision: str
    ai_recommendation: Optional[str] = None
    is_override: bool
    override_reason: Optional[str] = None
    rationale: str
    evidence_references: List[str]
    candidate_facing_status: str
    created_at: str


# ====================================================================
# 9. Multi-Brain Telemetry & Evaluation Schemas
# ====================================================================

class MultiBrainRunResponse(BaseModel):
    id: str
    organization_id: str
    role_name: str
    model_name: str
    prompt_version: str
    input_reference: str
    output_summary: Optional[str] = None
    duration_ms: int
    status: Literal["completed", "failed", "needs_review", "degraded_fallback"]
    error_details: Optional[str] = None
    created_at: str


class CandidateApplicationResponse(BaseModel):
    application_id: str
    job_id: Optional[str] = None
    job_title: str
    location: str
    status: str
    applied_at: str
